src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerOptionsExtensions.cs

```diff
@@ -16,9 +14,10 @@ internal static class AnalyzerOptionsExtensions
         private const string SeveritySuffix = "severity";
 
         private const string DotnetAnalyzerDiagnosticSeverityKey = DotnetAnalyzerDiagnosticPrefix + "." + SeveritySuffix;
+        private static readonly ConcurrentLruCache<string, string> s_categoryToSeverityKeyMap = new ConcurrentLruCache<string, string>(50);
 
         private static string GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(string category)
-            => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}";
+            => s_categoryToSeverityKeyMap.GetOrAdd(category, category, category => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}");
```

Comment:
Can the lambda be static?
```suggestion
            => s_categoryToSeverityKeyMap.GetOrAdd(category, category, static category => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}");
```

Reply:
Yup, sure. It is very important to make surr it doesn't capture.